Update README.md (#21)
ecmwf-lab/climetlab-s2s-ai-challenge: README.md

# S2S AI challenge Datasets

This is a `climetlab` plugin for Sub-seasonal to Seasonal (S2S) Artificial Intelligence Challenge: https://s2s-ai-challenge.github.io/.

In this README is a description of how to get the data for the S2S AI challenge. Here is a more general [description of the S2S data](https://confluence.ecmwf.int/display/S2S/Description). The data used for the S2S AI challenge is a subset of the S2S data library. More detail can be found at https://confluence.ecmwf.int/display/S2S  and https://confluence.ecmwf.int/display/S2S/Parameters.

There are several ways to use the datasets. Either by direct download (`wget`, `curl`, `browser`) for [`GRIB`](https://en.wikipedia.org/wiki/GRIB) and [NetCDF](https://en.wikipedia.org/wiki/NetCDF) formats; or using the `climetlab` python package with this addon, for `GRIB` and `NetCDF` and `zarr` formats. [`zarr`](https://zarr.readthedocs.io/en/stable/) is a cloud-friendly experimental data format and supports dowloading only the part of the data that is required. It has been designed to work better than classical format on a cloud environment (experimental).

# Installation

> `pip install climetlab climetlab_s2s_ai_challenge`

Author: ```suggestion
`pip install -U climetlab climetlab_s2s_ai_challenge`
```

> `pip install climetlab climetlab_s2s_ai_challenge`

Author: thanks

- `format`: data format, choose from [`netcdf` (always available), `grb` (only for `input-*`), `zarr` (experimental)]

## Coordinates

Overview of the time-related coordinates.

| `name` | CF convention `standard_name` | description | comment |
| --- | --- | --- | --- |
| `forecast_time` | forecast_reference_time | The forecast reference time in NWP is the "data time", the time of the analysis from which the forecast was made. It is not the time for which the forecast is valid. | |
| `lead_time` | forecast_period | Forecast period is the time interval between the forecast reference time and the validity time. |  |
| `valid_time` | time | time for which the forecast is valid | `forecast_time` + `lead_time` |

All datasets are on a global 1.5 degree grid.

## Parameter

`parameter` describes the variable to download. The most important two variables for the `s2s-ai-challenge` are the two target variables `t2m` and `tp`:

> | paramter | long_name | standard_name | unit | description & aggregation type | week 3-4 aggregation | week 5-6 aggregation | link to source |

Author: ```suggestion
| parameter | long_name | standard_name | unit | description & aggregation type | week 3-4 aggregation | week 5-6 aggregation | link to source |
```

| -------- | --------- | --- | --- | --- | --- | --- | --- |
| `t2m`    | 2m temperature | air_temperature | K | Temperature at 2m height averaged for the date given | average [day 14, day 27] | average [day 28, day 41] | [model](https://confluence.ecmwf.int/display/S2S/S2S+Surface+Air+Temperature), [observations](http://iridl.ldeo.columbia.edu/SOURCES/.NOAA/.NCEP/.CPC/.temperature/.daily/)|
| `tp`     | total precipitation | precipitation_amount | kg m-2 | Total precipitation accumulated from `forecast_time` until including `valid_time`, e.g. `lead_time = 1 days` accumulates precipitation_flux `pr` from 6-hourly steps 0,6,12,18 at date `forecast_time` | day 28 minus day 14 | day 42 minus day 28 | [model](https://confluence.ecmwf.int/display/S2S/S2S+Total+Precipitation) | 
| `pr`     | precipitation flux | precipitation_flux | kg m-2 | Precipitation accumulated for the date given | use `tp` | use `tp` | [observations](http://iridl.ldeo.columbia.edu/SOURCES/.NOAA/.NCEP/.CPC/.UNIFIED_PRCP/.GAUGE_BASED/.GLOBAL/.v1p0/.extREALTIME/.rain) |

> Given the different nature of the parameters, `tp` `lead_time=1 days` decribes conditions for the same day as `t2m` `lead_time=0 days`.

Reviewer: Please double-check with Frederic, or we could meet with him.
valid_time = forecast_time + lead_time.
This seems to mean that the tp and t2m data for the same value of valid_time do not refer to the physical state for same date, therefore are not valid for the same time.

Reviewer: decribes -> describes

Author: I wrote Frederic: forecast_reference_time Jan 2nd 2020, tp day 1 and t2m day 0 describe both the weather of Jan 2nd. tp as aggregated pr from Jan 2nd 00:00 to Jan 3rd 00:00 and t2m as the average of Jan 2nd.

He confirmed: Yes, this is correct. 2mtm and tp are indeed different types of variables. So, it will be challenging to treat the verification of both variables the same way

Reviewer: The sentence "Given the different nature of the parameters, `tp` `lead_time=1 days` decribes conditions for the same day as `t2m` `lead_time=0 days`." does not clarify, it only points to some tricky caveat but is not enough.

I would also add your nice example for jan 2nd.

Author: I do this in the renku datasets and inside the extra function. what do you mean by example? Written example or coding example?

Reviewer: Actually, I would give the example for day 2 and day 3 to emphasize that tp is accumulated.

> Given the different nature of the parameters, `tp` `lead_time=1 days` decribes conditions for the same day as `t2m` `lead_time=0 days`.

Author: ```suggestion
Given the different nature of the parameters, `tp` `lead_time=1 days` describes conditions for the same day (date of `forecast_time` 00:00 - 23:59) as `t2m` `lead_time=0 days`.
```

This terminology has lead to [some confusion in the past](https://en.wikipedia.org/wiki/Training,_validation,_and_test_sets#Confusion_in_terminology). 
Splitting the `hindcast-input` (`training-input`) dataset between training and validation is standard way and should be decided carefully.

> The `forecast-input` (`test-input`) must not be used as a validation dataset  : it must not be used to tune the hyperparameters or make decision about the ML model. 

Author: ```suggestion
The `forecast-input` (`test-input`) must not be used as a validation dataset  : it must not be used to tune the hyperparameters or make decision about the ML model. 
```

> Fostering discussions about how to prevent overfitting may be an outcome of the challenge.

Reviewer: Is it not true ?

Author: I didn't like the second sentence. We know what we need to do against overfitting. I will keep the first

Reviewer: ok

This dataset is available as `format`: `grib`, `netcdf` or `zarr`.
In this dataset, the data is available from 1999 for the oldest, to 2019/12/31 for the most recent. <!-- which model is available from 1998? -->
  - ECMWF hindcast data
    - `forecast_time`: from 2000/01/02 to 2019/12/31, corresponding to the weekly Thurdays in 2020.
    - `lead_time`: 0 to 46 days
    - `valid_time` (`forecast_time` + `lead_time`): from 2000/01/02 to 2020/02/13
    - availables parameters : `t2m(2t)/siconc(ci)/gh/lsm/msl/q/rsn/sm100/sm20/sp/sst/st100/st20/t/tcc/tcw/tp/ttr/u/v` (differing name in [MARS database](https://confluence.ecmwf.int/display/S2S/Parameters))
  - ECCC hindcast data 
    - `forecast_time`: from 2000/01/02 to 2019/12/31, corresponding to the weekly Thurdays in 2020.
    - `lead_time`: 1 to 32 days
    - `valid_time` (forecast_time + lead_time): from 
    - availables parameters: `t2m(2t)/siconc(ci)/gh/lsm/msl/q/rsn/sp/sst/t/tcc/tcw/tp/ttr/u/v` (differing name in [MARS database](https://confluence.ecmwf.int/display/S2S/Parameters))
    - parameters not available: `sm20,sm100,st20,st100`
  - NCEP hindcast data 

>     - `forecast_time` : from 1999/01/07 to 2010/12/30, corresponding to the weekly Thurdays in 2010.

Reviewer: Yes, it is better.
Maybe point that "Notice that the Thursday in 2010 do not happen the same days-of-years as the Thursdays in 2020", therefore the forecast_time dates in the NCEP data will not match the forecast_time in the ECCC and ECMWF data. For instance, 02.01.2009 is in the ECCC and ECMWF data (02.01.2020 is a Thurday) but 02.01.2009 is not in NCEP data (02.01.2010 is not a Thurday)".

Author: your explanation is right but I think a bit long to put it here.

Reviewer: fair point. As long as it is somewhere, that's fine.

hindcast_ncep.coords
Coordinates:
  * realization    (realization) int64 0 1 2 3
  * forecast_time  (forecast_time) datetime64[ns] 1999-01-07 ... 2010-01-07
  * lead_time      (lead_time) timedelta64[ns] 1 days 2 days ... 43 days 44 days
  * latitude       (latitude) float64 90.0 88.5 87.0 85.5 ... -87.0 -88.5 -90.0
  * longitude      (longitude) float64 0.0 1.5 3.0 4.5 ... 355.5 357.0 358.5
    valid_time     (forecast_time, lead_time) datetime64[ns] 1999-01-08 ... 2...
```

 List of files :
  [grib](https://storage.ecmwf.europeanweather.cloud/s2s-ai-challenge/data/training-input/0.3.0/grib/index.html),
  [netcdf](https://storage.ecmwf.europeanweather.cloud/s2s-ai-challenge/data/training-input/0.3.0/netcdf/index.html),
  zarr (not available)

### Forecast input (Test input)

### Forecast input

The `forecast-input` (`test-input`) dataset consists also in data from three different models: ECMWF (ecmf), ECCC (cwao), NCEP (eccc), for different dates.
These data are forecast data.
This could be used the input for applying the ML models in order to generate the output which is submitted for the challenge.

> Using data from earlier date that 2020/01/01 is also allowed during the prediction phase.

Reviewer: Is is documented elsewhere ?
If it is not allowed, maybe add "Using data from earlier date that 2020/01/01 is *not* allowed during the prediction phase."

Author: But should it be forbidden?

Author: IMO it is allowed to use a 3month prediction issued in October 2019 to predict Jan 2020

Reviewer: I agree, that's why I would keep the text "Using data from earlier date that 2020/01/01 is also allowed during the prediction phase". I don't understand why you removed it, except if it is documented elsewhere.

Author: I removed it because I think it is unnecessarily. furthermore, the rules are defined on the website. climetlab is just to get the data IMO.

- {datasetname} : In the URLs the dataset name must follow the ML naming (`training-input` or `training-output-reference` or `training-output-benchmark`).
- {format} is `netcdf`. Training output is also available as GRIB file,  using `format='grib'` and replacing `".nc"` by `".grib"`
- {fctype}: `hindcast` for training or `forecast` for test
- {parameter} is `t2m` for [surface temperature at 2m](https://confluence.ecmwf.int/display/S2S/S2S+Surface+Air+Temperature), `tp` for [total precipitation](https://confluence.ecmwf.int/display/S2S/S2S+Total+Precipitation)
- {origin} : `ecmwf` or `eccc` or `ncep`
- {weeks} from [`"34"`, `"56"`, `["34", "56"]`] only for `benchmark`

> - `YYYYMMDD` is the date of the 2020 forecast for `test-input`/`forecast-input`. The same dates are required for the on-the-fly `training-input`/`hindcast-input` but return the multi-year hindcast for that `MMDD`.

Reviewer: Btw, we could add the possibility to do this in the plugin code:
cml.load_dataset("training-input", date="0201")
cml.load_dataset("training-input", date=slice("0101","0201" )
cml.load_dataset("training-input", date="01*")

Author: Agree. Probably easier to understand.

Author: moved to issue. very low priority
